psycopg3/transaction.py

```diff
@@ -0,0 +1,86 @@
+"""
+Transaction context managers returned by Connection.transaction()
+"""
+
+# Copyright (C) 2020 The Psycopg Team
+
+# TODO: Should this module be _underscored since it's implementation detail?
+from types import TracebackType
+from typing import Optional, Type, TYPE_CHECKING
+
+from .pq import TransactionStatus
+
+if TYPE_CHECKING:
+    from .connection import Connection
+
+
+class Transaction:
+    def __init__(self, conn: "Connection") -> None:
+        self._conn = conn
+        self._outer_transaction = False
+        self._savepoint_name: Optional[str] = None
+        self._original_autocommit: bool
+
+    @property
+    def connection(self) -> "Connection":
+        return self._conn
+
+    def __enter__(self) -> None:
+        with self._conn.lock:
+            if self._conn.pgconn.transaction_status == TransactionStatus.IDLE:
+                self._outer_transaction = True
+                self._savepoint_name = None
+
+                self._original_autocommit = self._conn.autocommit
+                if self._conn._autocommit:
+                    self._conn._autocommit = False
+
+                self._conn._start_query()
```

Comment:
Uhm... are you doing the `_original_autocommit` here just because `_start_query()` would otherwise bail out? That `conn._start_query()` might actually be called `_maybe_begin()`, and here you are doing a job partially overlapping, by checking the `transaction_status`.

As on exit you are executing your COMMIT or ROLLBACK, you may as well execute a BEGIN here, instead of calling `_start_query()`.